cmd/dsl/main.go

```diff
@@ -81,34 +115,81 @@ func (p *Puppet) start(s Simulator, shim string) error {
 
 	cmd.Stderr = writer
 	cmd.Stdout = writer
-	err = cmd.Run()
+	// store cmd & logfile in puppet for use when we shut it down with e.g. the stop command
+	p.process = Process{cmd: cmd, logfile: logfile}
+	err = cmd.Start()
 	if err != nil {
 		return TestError{err: err, message: fmt.Sprintf("failure when creating puppet, see %s for information", filename)}
 	}
+
 	return nil
 }
 
-func (p Puppet) stop() {
+func (p *Puppet) stop() error {
+	// update the total message count before we stop this puppet
+	p.countMessages()
+	cmd, logfile := p.process.cmd, p.process.logfile
 	taplog(fmt.Sprintf("stopping %s (%s)", p.name, p.feedID))
-	p.stopProcess()
+	// issue an interrupt to the process (allows us to do cleanup in sbots)
+	// Windows doesn't support Interrupt
+	if runtime.GOOS == "windows" {
+		cmd.Process.Signal(os.Kill)
+	} else {
+		cmd.Process.Signal(os.Interrupt)
+	}
+
+	// last resort shutdown
+	go func() {
+		time.Sleep(2 * time.Second)
+		_ = cmd.Process.Signal(os.Kill)
+	}()
+
+	// wait for the process to wrap up
+	err := cmd.Wait()
+	if err != nil {
+		return TestError{err: err, message: fmt.Sprintf("failure when stopping puppet")}
```

Comment:
`fmt.Errorf`?

Reply:
didn't want to stray away from my established pattern, opting to do the TestError refactor in a separate PR :)